Mathlib/RingTheory/Torsion.lean

```diff
@@ -0,0 +1,47 @@
+/-
+Copyright (c) 2025 Yakov Pechersky. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yakov Pechersky
+-/
+import Mathlib.GroupTheory.Torsion
+import Mathlib.RingTheory.RootsOfUnity.Basic
+
+/-!
+# Torsion groups are the union of all subgroups of roots of unity
+
+-/
+
+variable (M : Type*) [CommMonoid M]
+
+lemma CommGroup.mem_torsion_iff_exists_mem_rootsOfUnity {x : Mˣ} :
+    x ∈ CommGroup.torsion Mˣ ↔ ∃ n ≠ 0, x ∈ rootsOfUnity n M := by
+  simp [CommGroup.mem_torsion, isOfFinOrder_iff_pow_eq_one, Nat.pos_iff_ne_zero]
+
+lemma rootsOfUnity_le_torsion (k : ℕ) [NeZero k] :
+    rootsOfUnity k M ≤ CommGroup.torsion Mˣ := by
+  intro
+  have : k ≠ 0 := NeZero.out
+  grind [CommGroup.mem_torsion_iff_exists_mem_rootsOfUnity]
+
+lemma CommGroup.torsion_eq_biSup_rootsOfUnity :
+    CommGroup.torsion Mˣ = ⨆ (n : ℕ) (_hn : n ≠ 0), rootsOfUnity n M := by
+  refine le_antisymm ?_ ?_
+  · intro x hx
+    rw [CommGroup.mem_torsion] at hx
+    -- without `(i := ...)`, timeout at `isDefEq`
+    rw [Subgroup.mem_biSup_of_directedOn (i := orderOf x) (orderOf_ne_zero_iff.mpr hx)]
```

Comment:
This sounds like it would benefit from the non-existent "IsDirectLimit" predicate...